### PR TITLE
rpc: bumpfee, improve doc for 'reduce_output' arg

### LINKED ISSUE #28180: Unsafe reduce_output when new coins are added
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behaviour

See inline discussion at https://github.com/bitcoin/bitcoin/pull/26467#discussion_r1269177446

@whitslack:

> Consider the case where I am paying someone on chain under the stipulation that _they_ will eat the mining fee. (Thus, I specify _their_ address in `subtractfeefrom` when I'm constructing my transaction.) Later, they tell me that they can't wait for confirmation any longer and need an urgent fee bump. If I naïvely specify their output as the `reduce_output`, I may end up paying them _**much more**_ than I intended to if I only have large UTxOs in my wallet.

### Expected behaviour

Not sure.

Improving the docs is one option, but this behaviour seem quite counter intuitive.

My own suggestion:

> Should we disable the adding of new inputs when `reduce_output` is set? If someone really intends to do that, they should probably use `outputs`.

### Steps to reproduce

Haven't tried.

### Relevant log output

_No response_

### How did you obtain Bitcoin Core

Compiled from source

### What version of Bitcoin Core are you using?

Master after #26467

### Operating system and version

Any

### Machine specifications

_No response_

### PR DESCRIPTION
**Pull request by furszy**
Fixes #28180. Resulted from discussions with S3RK, achow101, and Murch.

The current argument name and description are dangerous as it don't
describe the case where the user selects the recipient output as the
change address. This one could end up been increased by the inputs
minus outputs remainder. Which, when `bumpfee` adds new inputs
to the transaction, leads the process to send more coins to the
recipient. Which is not what the user would expect from a
'reduce_output' param naming.

### CODE REVIEW

File: test/functional/wallet_bumpfee.py [hunk 1]
```diff
@@ -283,10 +283,10 @@ def test_single_output(self):
         tx = wallet.sendall(recipients=[wallet.getnewaddress()], fee_rate=2, options={"inputs": [utxos[0]]})
 
         # Reduce the only output with a crazy high feerate, should fail as the output would be dust
```
Comment (kevkevinpal) [inline test/functional/wallet_bumpfee.py:285]: We might want to change this comment to
"Change the output index to the only output with a crazy high feerate, should fail as the output would be dust"

File: test/functional/wallet_bumpfee.py [hunk 2]
```diff
@@ -283,10 +283,10 @@ def test_single_output(self):
...
-        assert_raises_rpc_error(-4, "The transaction amount is too small to pay the fee", wallet.bumpfee, txid=tx["txid"], options={"fee_rate": 1100, "reduce_output": 0})
+        assert_raises_rpc_error(-4, "The transaction amount is too small to pay the fee", wallet.bumpfee, txid=tx["txid"], options={"fee_rate": 1100, "change_output_index": 0})
 
         # Reduce the only output successfully
```
Comment (kevkevinpal) [inline test/functional/wallet_bumpfee.py:288]: We might want to change this comment to
"Change the output index successfully"

File: src/wallet/feebumper.cpp [hunk 3]
```diff
@@ -162,10 +162,10 @@ bool TransactionCanBeBumped(const CWallet& wallet, const uint256& txid)
 }
 
 Result CreateRateBumpTransaction(CWallet& wallet, const uint256& txid, const CCoinControl& coin_control, std::vector<bilingual_str>& errors,
-                                 CAmount& old_fee, CAmount& new_fee, CMutableTransaction& mtx, bool require_mine, const std::vector<CTxOut>& outputs, std::optional<uint32_t> reduce_output)
+                                 CAmount& old_fee, CAmount& new_fee, CMutableTransaction& mtx, bool require_mine, const std::vector<CTxOut>& outputs, std::optional<uint32_t> change_output_index)
 {
     // Cannot both specify new outputs and an output to reduce
```
Comment (kevkevinpal) [inline src/wallet/feebumper.cpp:167]: Might want to change to
"Cannot both specify new outputs and an output to change"

Comment (furszy) [reply src/wallet/feebumper.cpp]: done, thx

File: test/functional/wallet_bumpfee.py (hunk 1 above)
Comment (furszy) [reply test/functional/wallet_bumpfee.py]: done, thx

File: test/functional/wallet_bumpfee.py (hunk 2 above)
Comment (furszy) [reply test/functional/wallet_bumpfee.py]: done, thx

**Comment by S3RK** [comment]
Approach ACK

Explicitly specifying change output for bumpfee and reducing an arbitrary output is two different features.
#26467 originally was developed to specify the change output, hence updating documentation is the correct way to fix #28180 for 26.0

We can discuss whether we want to support deducting fees from an arbitrary output in a separate issue.

File: src/wallet/feebumper.cpp [hunk 4]
```diff
@@ -162,11 +162,11 @@ bool TransactionCanBeBumped(const CWallet& wallet, const uint256& txid)
...
 Result CreateRateBumpTransaction(CWallet& wallet, const uint256& txid, const CCoinControl& coin_control, std::vector<bilingual_str>& errors,
-                                 CAmount& old_fee, CAmount& new_fee, CMutableTransaction& mtx, bool require_mine, const std::vector<CTxOut>& outputs, std::optional<uint32_t> reduce_output)
+                                 CAmount& old_fee, CAmount& new_fee, CMutableTransaction& mtx, bool require_mine, const std::vector<CTxOut>& outputs, std::optional<uint32_t> change_output_index)
 {
-    // Cannot both specify new outputs and an output to reduce
-    if (!outputs.empty() && reduce_output.has_value()) {
-        errors.push_back(Untranslated("Cannot specify both new outputs to use and an output index to reduce"));
+    // For now, cannot specify both new outputs to use and an output index to send change
+    if (!outputs.empty() && change_output_index.has_value()) {
+        errors.push_back(Untranslated("Cannot specify both new outputs to use and an output index to send change"));
```
Comment (murchandamus) [inline src/wallet/feebumper.cpp:169]:
```suggestion
        errors.push_back(Untranslated("The options 'outputs' and 'change_output_index' are incompatible. You can only either specify a new set of outputs, or designate a change output to be recycled."));
```

**Review by murchandamus — COMMENTED** [review]
Approach ACK, I do have a few suggestions on the documentation and naming of things, though.

File: src/rpc/client.cpp [hunk 5]
```diff
@@ -263,13 +263,13 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "bumpfee", 1, "fee_rate"},
     { "bumpfee", 1, "replaceable"},
     { "bumpfee", 1, "outputs"},
-    { "bumpfee", 1, "reduce_output"},
+    { "bumpfee", 1, "change_output_index"},
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/rpc/client.cpp]: sounds good.

File: src/wallet/rpc/spend.cpp [hunk 6]
```diff
@@ -1016,10 +1016,10 @@ static RPCHelpMan bumpfee_helper(std::string method_name)
...
                              "At least one output of either type must be specified.\n"
-                             "Cannot be provided if 'reduce_output' is specified.",
+                             "Cannot be provided if 'change_output_index' is specified.",
                         OutputsDoc(),
                         RPCArgOptions{.skip_type_check = true}},
-                    {"reduce_output", RPCArg::Type::NUM, RPCArg::DefaultHint{"not set, detect change automatically"}, "The 0-based index of the output from which the additional fees will be deducted. In general, this should be the position of change output. Cannot be provided if 'outputs' is specified."},
+                    {"change_output_index", RPCArg::Type::NUM, RPCArg::DefaultHint{"not set, detect change automatically"}, "The change of the bumped transaction will be sent to the output script specified by this 0-based index parameter. The to-be-replaced transaction output pointed by this parameter could either be increased or decreased depending on the inputs minus outputs remainder. Cannot be provided if 'outputs' is specified."},
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/wallet/rpc/spend.cpp]: sure, done.

File: src/wallet/feebumper.cpp (hunk 4 above)
Comment (furszy) [reply src/wallet/feebumper.cpp]: done

**Comment by ismaelsadeeq** [comment]
Concept ACK

**Review by murchandamus — APPROVED** [review]
ACK b3db8c9d5ccfe5c31341169fa7ac044427122921

**Comment by S3RK** [comment]
ACK b3db8c9d5ccfe5c31341169fa7ac044427122921

Maybe add a test case where bumping leads to new input and increased change output?

**Comment by achow101** [comment]
ACK b3db8c9d5ccfe5c31341169fa7ac044427122921

**Comment by luke-jr** [comment]
The bug here is the behaviour, not the documentation! It *shouldn't* do that!

**Comment by achow101** [comment]
> The bug here is the behaviour, not the documentation! It _shouldn't_ do that!

Changing the documentation to reflect the current behavior does not preclude reintroducing `reduce_output` in the future that has the correct behavior. The discussion at CoreDev concluded that both an option to indicate which output is change and an option to indicate which output the fee should be taken from are reasonable things that a user may wish to do. As the former is what was implemented, the documentation should reflect that.

**Pull request merged**